### ISSUE TITLE
dll

### ISSUE THREAD

**Issue by mbazzo** [TeraHz/PCA9685#4]
friend,
do you have the dll of this code so I can use it in C #? I do not handle anything in C ++.
Thankful.

**Comment by TeraHz** [comment]
Hi @mbazzo 
The code is fairly unix specific, I don't think it will work on Windows.
You can, however, write C# on linux, if I'm not mistaken, via MONO. 
Good luck!

**Comment by mbazzo** [comment]
Yes Yes...
I'm doing a C # application to run on a raspberry. (Raspbian).
I do not know how to generate the dll to be able to import into my code.

**Comment by TeraHz** [comment]
I really know nothing about C#/Dlls, but a quick google turned out this: http://www.mono-project.com/docs/advanced/pinvoke/ It looks like you can just import the .so produced by the makefile. You should be able to run make (after updating the paths to your compiler) and that will produce a .so file for you to import. 
You'll also need the I2C library that goes with it https://github.com/TeraHz/I2C

**Comment by mbazzo** [comment]
Right...
Can you just give me some tips on how to use your lib?
parameters that you use example

bus = 0x40
address = 0

it is?

**Comment by mbazzo** [comment]
I'm doing this:

internal class PCA9685Helper
    {
        internal const string PCA9685Library = "libPCA9685.so.0.1";

        [DllImport(PCA9685Library, EntryPoint = "PCA9685", SetLastError = true)]
        public static extern void PCA9685(int bus, int address);

        [DllImport(PCA9685Library, EntryPoint = "reset", SetLastError = true)]
        public static extern void Reset();

        [DllImport(PCA9685Library, EntryPoint = "setPWMFreq", SetLastError = true)]
        public static extern void SetPWMFreq(int channel, int value);

        [DllImport(PCA9685Library, EntryPoint = "setPWM", SetLastError = true)]
        public static extern void SetPWM(int channel, int start, int end);

        [DllImport(PCA9685Library, EntryPoint = "getPWM", SetLastError = true)]
        public static extern void GetPWM(int channel);
    }

**Comment by TeraHz** [comment]
That looks correct. Seems like you got it.

**Comment by mbazzo** [comment]
I'm just thinking that I did not generate the .so file correctly. It's giving him an error.

I just need to run "make", right?

generating the file libPCA9685.so.0.1

**Comment by TeraHz** [comment]
Yes, as long as the CXX variable in the Makefile is set to a working compiler, just typing `make` should produce a .so file. If you're getting a .so file after running `make` then you're all set. 

What is the error?

**Comment by mbazzo** [comment]
System.EntryPointNotFoundException: Unable to find an entry point named 'PCA9685' in DLL 'libPCA9685.so.0.1'.

**Comment by TuxLeon** [comment]
Could it bee that your problem is related to C++ code. The PCA9685 is C++ code and your DLLimport declaration looks like C-Code. When I understand this instructions http://www.mono-project.com/docs/advanced/pinvoke/ right then you need to write some kind of wrapper for the C++ class to use it in your DLLimport.
Or you provide the correct  "Entrypoint". In my case I is _ZTV7PCA9685 (I guess) and not PCA9685 as you specified in your DLLimport. You could get this value by the following command
objdump -T <your library>.

**Comment by mbazzo** [comment]
I think I got it ... I'm calling the wrong entry point.
You have the list of Entry Points, because this is giving me a huge list.

**Comment by mbazzo** [comment]
Hello my friend, how are you?
How to draw the symbol '_ZN7PCA9685' from the methods.
I still can not make it work inside C #.

If you can help me.
Hugs
